### DISCUSSION TITLE
Is watching providers conditionally valid?

### DISCUSSION THREAD

**Discussion by mono0926** [rrousselGit/riverpod#711] [Q&A] [answered]
Is this kind of code valid or not?

```dart
final isWatch = StateProvider((ref) => true);

final provider1 = Provider.autoDispose((ref) {
  return 'hello';
});

final provider2 = Provider.autoDispose((ref) {
  if (ref.watch(isWatch).state) {
    return ref.watch(provider1); // Is this valid?
  }
  return 'world';
});
```

Should be written like this?

```dart
final provider2 = Provider.autoDispose((ref) {
  final watch = ref.watch(isWatch).state;
  final p1 = ref.watch(provider1);
  if (watch) {
    return p1;
  }
  return 'world';
});
```

If so, is it mentioned in the documentation?

As a similar thing, there used to be a note in the [flutter_hooks](https://github.com/rrousselGit/flutter_hooks) documentation that said to put `useXxx()` at the beginning of the scope, but flutter_hooks seems to have lost that restriction and the note were removed(https://github.com/rrousselGit/flutter_hooks/commit/ecbffc75c6280e9ec1992f5b9360c9b89b018869#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5).
I would like to know which is the case with riverpod.

**Comment by rrousselGit** [comment, marked as answer]
Eh the fact that the hooks Readme lost the mention is incorrect. Nothing changed there


But yes, conditionally depending on providers is valid with Riverpod.
This was one of the design decisions

**Reply by rrousselGit** [reply]
One thing worth nothing:

While there's no particular limitation about this at the moment, I would advise against any non-static usage of a provider
As in, don't pass a provider as parameter to a function


With static meta programming, I'm looking into compile-time optimisations and various improvements which would add this restriction.

**Reply by mono0926** [reply]
@rrousselGit 

Thanks for the answer!

But, I've found a weird behavior.
It seems that `autoDispose` doesn't work correctly if `watch`ed conditionally 🤔 

1. Run the app
2. Press `watch: true` button, and it will change to `watch: false`.
3. Press `show: true` button sometimes, and `provider2` will be disposed correctly but `provider1` won't be disposed at all.

log:

```
I/flutter ( 6811): init 2
I/flutter ( 6811): init 1
I/flutter ( 6811): dispose 2
I/flutter ( 6811): init 2
I/flutter ( 6811): dispose 2
I/flutter ( 6811): init 2
I/flutter ( 6811): dispose 2
I/flutter ( 6811): init 2
```

```dart
import 'package:flutter/material.dart';
import 'package:hooks_riverpod/hooks_riverpod.dart';

void main() => runApp(
      const ProviderScope(
        child: App(),
      ),
    );

class App extends ConsumerWidget {
  const App({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    return const MaterialApp(
      home: HomePage(),
    );
  }
}

final isShow = StateProvider((ref) => true);
final isWatch = StateProvider((ref) => true);

final provider1 = Provider.autoDispose((ref) {
  print('init 1');
  ref.onDispose(() {
    print('dispose 1');
  });
  return 'hello';
});

final provider2 = Provider.autoDispose((ref) {
  print('init 2');
  ref.onDispose(() {
    print('dispose 2');
  });
  if (ref.watch(isWatch).state) {
    return ref.watch(provider1);
  }
  return 'world';
});

class HomePage extends ConsumerWidget {
  const HomePage({Key? key}) : super(key: key);
  @override
  Widget build(BuildContext context, WidgetRef ref) {
    return Scaffold(
      appBar: AppBar(),
      body: Center(
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            if (ref.watch(isShow).state) const Center(child: _Body()),
            ElevatedButton(
              child: Text('show: ${ref.watch(isShow).state}'),
              onPressed: () => ref.read(isShow).state = !ref.read(isShow).state,
            ),
            ElevatedButton(
              child: Text('watch: ${ref.watch(isWatch).state}'),
              onPressed: () =>
                  ref.read(isWatch).state = !ref.read(isWatch).state,
            ),
          ],
        ),
      ),
    );
  }
}

class _Body extends ConsumerWidget {
  const _Body({Key? key}) : super(key: key);
  @override
  Widget build(BuildContext context, WidgetRef ref) {
    return Text(ref.watch(provider2));
  }
}
```

Riverpod version is `1.0.0-dev.7`.

**Reply by rrousselGit** [reply]
That could be a bug. Do you mind raising an issue?

**Reply by mono0926** [reply]
@rrousselGit 

Okay, I raised an issue: https://github.com/rrousselGit/river_pod/issues/712

My question have been solved, thanks 👍

**Reply by mono0926** [reply]
> Eh the fact that the hooks Readme lost the mention is incorrect. Nothing changed there

And, I've risen the issue about it: https://github.com/rrousselGit/flutter_hooks/issues/262

**Comment by serbixote** [comment]
Is something like this still valid?

```dart
@override
Widget build(BuildContext context) {
    if (myCondition) {
        ref.watch(myProvider);
    }
}
```
Is there any other way to start/stop watching providers during run time??

**Reply by AhmedLSayed9** [reply]
I believe that's fine.

**Reply by rrousselGit** [reply]
Being able to conditionally watch providers is expected

**Reply by serbixote** [reply]
Thanks for the reply.

Is it anti-pattern to create providers dynamically??

I have a provider with the .family modifier and I want to use it within a for loop with X different parameter in the build method. Is it okey?

Would that make my widget be listening X different providers? (Which would be what I desire)

Since it is a stream provider, would this make to recreate the web-socket every time the build method is called?
